src/hotspot/share/cds/aotClassLinker.cpp

```diff
@@ -0,0 +1,293 @@
+ *
+ */
+
+#include "precompiled.hpp"
+#include "cds/aotClassLinker.hpp"
+#include "cds/aotConstantPoolResolver.hpp"
+#include "cds/aotLinkedClassTable.hpp"
+#include "cds/archiveBuilder.hpp"
+#include "cds/archiveUtils.inline.hpp"
+#include "cds/cdsConfig.hpp"
+#include "cds/heapShared.hpp"
+#include "cds/lambdaFormInvokers.inline.hpp"
+#include "classfile/classLoader.hpp"
+#include "classfile/dictionary.hpp"
+#include "classfile/systemDictionary.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "classfile/vmClasses.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/constantPool.inline.hpp"
+#include "oops/instanceKlass.hpp"
+#include "oops/klass.inline.hpp"
+#include "runtime/handles.inline.hpp"
+
+AOTClassLinker::ClassesTable* AOTClassLinker::_vm_classes = nullptr;
+AOTClassLinker::ClassesTable* AOTClassLinker::_candidates = nullptr;
+GrowableArrayCHeap<InstanceKlass*, mtClassShared>* AOTClassLinker::_sorted_candidates = nullptr;
+
+#ifdef ASSERT
+bool AOTClassLinker::is_initialized() {
+  assert(CDSConfig::is_dumping_archive(), "AOTClassLinker is for CDS dumping only");
+  return _vm_classes != nullptr;
+}
+#endif
+
+void AOTClassLinker::initialize() {
+  assert(!is_initialized(), "sanity");
+
+  _vm_classes = new (mtClass)ClassesTable();
+  _candidates = new (mtClass)ClassesTable();
+  _sorted_candidates = new GrowableArrayCHeap<InstanceKlass*, mtClassShared>(1000);
+
+  for (auto id : EnumRange<vmClassID>{}) {
+    add_vm_class(vmClasses::klass_at(id));
+  }
+
+  assert(is_initialized(), "sanity");
+
+  AOTConstantPoolResolver::initialize();
+}
+
+void AOTClassLinker::dispose() {
+  assert(is_initialized(), "sanity");
+
+  delete _vm_classes;
+  delete _candidates;
+  delete _sorted_candidates;
+  _vm_classes = nullptr;
+  _candidates = nullptr;
+  _sorted_candidates = nullptr;
+
+  assert(!is_initialized(), "sanity");
+
+  AOTConstantPoolResolver::dispose();
+}
+
+bool AOTClassLinker::is_vm_class(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+  return (_vm_classes->get(ik) != nullptr);
+}
+
+void AOTClassLinker::add_vm_class(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+  bool created;
+  _vm_classes->put_if_absent(ik, &created);
+  if (created) {
+    add_candidate(ik);
+    InstanceKlass* super = ik->java_super();
+    if (super != nullptr) {
+      add_vm_class(super);
+    }
+    Array<InstanceKlass*>* ifs = ik->local_interfaces();
+    for (int i = 0; i < ifs->length(); i++) {
+      add_vm_class(ifs->at(i));
+    }
+  }
+}
+
+bool AOTClassLinker::is_candidate(InstanceKlass* ik) {
+  return (_candidates->get(ik) != nullptr);
+}
+
+void AOTClassLinker::add_candidate(InstanceKlass* ik) {
+  _candidates->put_when_absent(ik, true);
+  _sorted_candidates->append(ik);
+}
+
+bool AOTClassLinker::try_add_candidate(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+  assert(CDSConfig::is_dumping_aot_linked_classes(), "sanity");
+
+  if (!SystemDictionaryShared::is_builtin(ik)) {
```

Comment:
What does this actually mean by "built-in"?

Reply:
Boot/Platform/App loaders. The meaning of "built-in" is documented at the top of systemDictionaryShared.hpp